### PR TITLE
Tigers - Kat Hawkins

### PR DESCRIPTION
**Pull request by kathawkins**

### CODE REVIEW

File: viewing_party/party.py [hunk 1]
```diff
@@ -1,23 +1,213 @@
 # ------------- WAVE 1 --------------------
 
-def create_movie(title, genre, rating):
-    pass
+from symbol import single_input
```
Comment (spitsfire) [inline viewing_party/party.py:3]: doesn't look like we are using this package anywhere. sometimes VSCode will try to be helpful and auto-import packages it thinks we are trying to type out. let's get rid of this

```suggestion

```

File: viewing_party/party.py [hunk 2]
```diff
@@ -1,23 +1,213 @@
...
 
-def create_movie(title, genre, rating):
-    pass
+from symbol import single_input
+
+
+def create_movie(title, genre, rating): #Wave 1, step 1 done
+    movie_dict={}
+    if title is None or genre is None or rating is None:
+        return None
```
Comment (spitsfire) [inline viewing_party/party.py:9]: Think about putting this at the top of the function first. It saves us from creating space in memory, defining variables, etc. that we are simply going to throw away because the parameters didn't meet our rules, ie, `movie_dict`

File: viewing_party/party.py [hunk 3]
```diff
@@ -1,23 +1,213 @@
...
+
+    movie_dict["title"]=title
+    movie_dict["genre"]=genre
+    movie_dict["rating"]=rating
```
Comment (spitsfire) [inline viewing_party/party.py:13]: this works just fine! We could also do this all at once: 
```suggestion
    movie_dict = {
        "title": title
        "genre": genre
        "rating": rating
    }

```
Now, obviously this isn't one line like this, so we aren't saving space. But we can do it all in one expression instead of 4.

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,213 @@
...
+    return movie_dict
+
+def add_to_watched(user_watched_dict,movie_info):
+    user_watched_dict["watched"] = [movie_info]
+    return user_watched_dict
```
Comment (spitsfire) [inline viewing_party/party.py:18]: careful here! You are re-assigning `user_data["watched"]` value, *not* appending *to* the list already inside. 

So, what if this same user added another movie to "watched"? It would overwrite the first movie with the new movie, instead of *appending it* to the list already declared in `user_data["watched"]`

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,213 @@
...
+
+def add_to_watchlist(user_watchlist_dict, movie_info):
+    user_watchlist_dict["watchlist"] = [movie_info]
+    return user_watchlist_dict
```
Comment (spitsfire) [inline viewing_party/party.py:22]: same as above! we aren't adding the movie *to* the user's data here. We are simply overwriting the original list.

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,213 @@
...
+
+def watch_movie(single_user_data,movie_title):    
+    list_of_watchlist_movie_info=single_user_data["watchlist"]
+    list_of_watched_movie_info=single_user_data["watched"]
+
+    #Adds movie info to list of watched movies if it is in the watchlist and then removes the movie info from the watchlist
+    for index,movie_info in enumerate(list_of_watchlist_movie_info):
+        if movie_title == movie_info["title"]:
+            list_of_watched_movie_info.append(movie_info)
+            list_of_watchlist_movie_info.pop(index)
```
Comment (spitsfire) [inline viewing_party/party.py:32]: nicely done! I see you used `enumerate` in order to access index position and the movie dictionary more easily. 

We could use `remove` method instead of `pop` if we wanted to use a simpler for loop, which takes a value as a parameter instead of an index position.

Your solution works just fine as is!

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,213 @@
...
+    list_of_watched_movie_info=single_user_data["watched"]
+    
+    #Calculates the number of watched movies
+    number_of_watched_movies=len(list_of_watched_movie_info)
+
+    list_of_movie_ratings=[]
+    sum_of_all_ratings=0
+
+    if len(list_of_watched_movie_info) == 0:
+        return 0.0
```
Comment (spitsfire) [inline viewing_party/party.py:50]: Let's move this to the top, so we don't have to make all these variables first, and then find out this function needs to return early and throw out all those variables we just made.

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,213 @@
...
+
+    #Adds rating of each watched movie to a list
+    for movie_info in list_of_watched_movie_info:
+        list_of_movie_ratings.append(movie_info["rating"])
+
+    #Determines the sum of all ratings
+    for rating in list_of_movie_ratings:
+        sum_of_all_ratings+=rating
```
Comment (spitsfire) [inline viewing_party/party.py:58]: How could we combine these two for loops? We are already looping over each movie on line 53, why not add its rating to `sum_of_all_ratings` right there instead of in a separate for loop?

This would also save us from creating `movie_ratings` altogether

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,213 @@
...
+    return average_rating
+
+def get_most_watched_genre(single_user_data):
+    list_of_watched_movie_info=single_user_data["watched"]
+    list_of_genres=[]
+    genre_counter_dict={}
+    list_of_watched_genre_counts=[]
+
+    if len(list_of_watched_movie_info) == 0:
+        return None
```
Comment (spitsfire) [inline viewing_party/party.py:71]: same as above! let's put this at the top

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,213 @@
...
+    #Adds genre of each watched movie to a list
+    for movie_info in list_of_watched_movie_info:
+        list_of_genres.append(movie_info['genre'])
+
+    #Adds genre as the key to a dictionary with a value that counts the number of times it is in the genre list
+    for genre in list_of_genres:
+        if not genre in genre_counter_dict:
+            genre_counter_dict[genre] = 1
+        else:
+            genre_counter_dict[genre] += 1
```
Comment (spitsfire) [inline viewing_party/party.py:82]: love this! great job! how could we combine these two for loops together? We are already looping through each movie on line 74, why not count the genres while we are in the first loop.

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,213 @@
...
+    for genre_count_value in genre_counter_dict.values():
+        list_of_watched_genre_counts.append(genre_count_value)
+
+    #Determines the highest genre count value
+    max_watched_genre_count=max(list_of_watched_genre_counts)
+
+    #Returns a genre from the genre:counter dictionary with a counter value that is the same as the highest genre count value
+    for genre,genre_count_value in genre_counter_dict.items():
+        if genre_count_value == max_watched_genre_count:
+            return genre
```
Comment (spitsfire) [inline viewing_party/party.py:94]: ohhh ok interesting approach! gotta ask ;) food for thought: how could we combine these two for loops to together and find the max?

There is a for loop of some kind being run under the hood of the `max` function, so it's not necessarily saving us time or a for loop on its own

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,213 @@
...
+    for user_movie_info in list_of_user_movie_info:
+        list_of_user_movie_titles.append(user_movie_info["title"])
+ 
+    #Adds friends' watched movie titles to a list
+    for frnds_watched_movie_info in list_of_frnds_watched_dicts:
+        for frnds_watched_movie_titles in frnds_watched_movie_info["watched"]:
+            list_of_friends_movie_titles.append(frnds_watched_movie_titles["title"])
+
+    #Uses sets data structure and difference method to add user's unique move titles that only the user has seen to a list
+    list_of_unique_user_movie_titles=set(list_of_user_movie_titles).difference(set(list_of_friends_movie_titles))
```
Comment (spitsfire) [inline viewing_party/party.py:117]: nice job using sets to help filter out duplicates! we could use them immediately on lines 103 and 104 by setting `list_of_friends_movie_titles=()`, then loop over the user's movies

Any time it tries to add a value, it has to check if it's already in there. If it is, it won't add it. 

Your solution works just fine, though! something to simply consider

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,213 @@
...
+    for user_movie_info in list_of_user_movie_info:
+        list_of_user_movie_titles.append(user_movie_info["title"])
+
+    #Adds friends' watched movie titles to a list
+    for frnds_watched_movie_info in list_of_frnds_watched_dicts:
+        for frnds_watched_movie_titles in frnds_watched_movie_info["watched"]:
+            list_of_friends_movie_titles.append(frnds_watched_movie_titles["title"])
+    
+    #Uses sets data structure and difference method to add friends' unique move titles that at least one friend has seen to a list
+    list_of_unique_frnds_movie_titles=set(list_of_friends_movie_titles).difference(set(list_of_user_movie_titles))
```
Comment (spitsfire) [inline viewing_party/party.py:142]: hmm where have we seen this code already? maybe we could turn this into a helper function that we call upon inside this function.

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,213 @@
...
 # -----------------------------------------
 
+def get_available_recs(single_user_data):
+    list_of_user_movie_info=single_user_data["watched"]
+    list_of_user_movie_hosts=single_user_data["subscriptions"]
+    list_of_users_unwatched_movie_info=[]
+    list_of_recommended_movie_info=[]
+
+    #Uses wave_03 function to add user's friends' unique watched movie info to a list
+    list_of_unique_frnds_movie_info=get_friends_unique_watched(single_user_data)
```
Comment (spitsfire) [inline viewing_party/party.py:162]: nice job reusing functions!

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,213 @@
...
 # -----------------------------------------
 
+def get_new_rec_by_genre(single_user_data):
+    #Returns an empty list if the user does not have a most frequent genre
+    if len(single_user_data["watched"]) == 0:
+        return []
+    
+    #Returns an empty list if the user's friends do not have watchlists with movie info that can be used to get available movie recommendations
+    if len(get_friends_unique_watched(single_user_data)) == 0 :
+        return []
```
Comment (spitsfire) [inline viewing_party/party.py:186]: great job putting these edge case checks at the top (sometimes called "guard clauses")

**Review by spitsfire — COMMENTED** [review]
Nicely done, Kat! Great job passing all tests.

I've added some feedback throughout your project to point out some places we can refactor in order to follow common conventions and standard practices when it comes to Python.

Things to consider as you work on future projects:
- let's turn repeat code, like filtering out duplicate movies, into helper functions, so that we can write the code once and only once!
- your variables were very descriptive! but also *very* long. Sometimes we have to balance our desire to be descriptive but also readable for other people. Consider shortening some of these variables a bit
